### PR TITLE
Rewrite

### PR DESCRIPTION
**Pull request by avik-pal**

### CODE REVIEW

File: src/core.jl [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+### initialparameters(rng, l) --> Must return a NamedTuple/ComponentArray
+### initialstates(rng, l)     --> Must return a NamedTuple
+### parameterlength(l)        --> Integer
+### statelength(l)            --> Integer
+### l(x, ps, st)
+
+abstract type AbstractExplicitLayer end
+
+initialparameters(::AbstractRNG, ::Any) = ComponentArray{Float32}()
+initialparameters(l) = initialparameters(Random.GLOBAL_RNG, l)
```
[the remark this answers is not in the document]
Comment (avik-pal) [reply src/core.jl:12]: I am down to completely remove this.

File: src/layers/basic.jl [hunk 2]
```diff
@@ -0,0 +1,361 @@
+"""
+    ReshapeLayer(dims)
+
+Reshapes the passed array to have a size of `(dims..., :)`
+"""
+struct ReshapeLayer{N} <: AbstractExplicitLayer
+    dims::NTuple{N,Int}
+end
+
+Base.@pure function (r::ReshapeLayer)(x::AbstractArray, ps, st::NamedTuple)
```
[the remark this answers is not in the document]
Comment (avik-pal) [reply src/layers/basic.jl:10]: Yes I see. I will remove these.

File: src/utils.jl [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+@inline _norm(x; dims=Colon()) = sqrt.(sum(abs2, x; dims=dims))
+
+# Compute norm over all dimensions except `except_dim`
+@inline _norm_except(x::AbstractArray{T,N}, except_dim) where {T,N} = _norm(x; dims=filter(i -> i != except_dim, 1:N))
+@inline _norm_except(x::AbstractArray{T,N}) where {T,N} = _norm_except(x, N)
+
+# Handling ComponentArrays
+gpu(c::ComponentArray) = ComponentArray(gpu(getdata(c)), getaxes(c))
+cpu(c::ComponentArray) = ComponentArray(cpu(getdata(c)), getaxes(c))
```
[the remark this answers is not in the document]
Comment (avik-pal) [reply src/utils.jl:29]: Most of these ComponentArrays piracies will have to be upstreamed. The main issue with not defining this is the default recursively goes through the properties which creates issues with the construction.

File: src/layers/basic.jl [hunk 4]
```diff
@@ -0,0 +1,447 @@
...
+    end
+end
+
+parameterlength(d::Dense{bias}) where {bias} = bias ? d.out_dims * (d.in_dims + 1) : d.out_dims * d.in_dims
+statelength(d::Dense) = 0
+
+Base.@pure function (d::Dense{bias,λT})(
+    x::AbstractArray{T,N}, ps::Union{ComponentArray,NamedTuple}, st::NamedTuple
+) where {bias,T,N,λT}
+    if bias
```
[the remark this answers is not in the document]
Comment (avik-pal) [reply src/layers/basic.jl:387]: Won't these get compiled away. bias is part of the type information so it should dispatch on that, right?

[the remark this answers is not in the document]
Comment (ToucheSir) [reply src/layers/basic.jl:387]: Not sure, I think `@code_adjoint` would tell you?

[the remark this answers is not in the document]
Comment (avik-pal) [reply src/layers/basic.jl:387]: Yeah, julia compiles these things away but Zygote seems to keep them around

File: src/layers/normalize.jl [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+                μ2,
+                σ²2,
+                BN.momentum;
+                eps=BN.ϵ,
+                training=istraining(st),
+            ),
+        )
+    if track_stats
+        @set! st.μ = μ2
+        @set! st.σ² = σ²2
```
[the remark this answers is not in the document]
Comment (avik-pal) [reply src/layers/normalize.jl:136]: No. I just keep forgetting there is a merge function

File: src/layers/basic.jl (hunk 4 above)
[the remark this answers is not in the document]
Comment (ToucheSir) [reply src/layers/basic.jl:388]: Ok, unfortunately that's what I expected. Zygote works on untyped, unoptimized IR and I assume this DCE + constprop runs after that step.

File: src/autodiff.jl [hunk 6]
```diff
@@ -0,0 +1,50 @@
+Base.zero(s::NamedTuple{(),Tuple{}}) = s
+
+Base.zero(::Symbol) = Symbol()
+
+Base.zero(nt::NamedTuple{fields}) where {fields} = NamedTuple{fields}(zero.(values(nt)))
```
[the remark this answers is not in the document]
Comment (avik-pal) [reply src/autodiff.jl:5]: Added this while toying around with Yota.jl. Needs to be removed from this PR

File: src/core.jl [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+statelength(l::AbstractExplicitLayer) = statelength(initialstates(l))
+statelength(nt::Union{NamedTuple,Tuple}) = length(nt) == 0 ? 0 : sum(statelength, nt)
+statelength(a::AbstractArray) = length(a)
+statelength(x::Union{Number,Symbol}) = 1
+statelength(x) = 0
+
+setup(rng::AbstractRNG, l::AbstractExplicitLayer) = (initialparameters(rng, l), initialstates(rng, l))
+setup(l::AbstractExplicitLayer) = setup(Random.GLOBAL_RNG, l)
+
+apply(model::AbstractExplicitLayer, x, ps::Union{ComponentArray,NamedTuple}, st::NamedTuple) = model(x, ps, st)
```
[the remark this answers is not in the document]
Comment (avik-pal) [reply src/core.jl:30]: I might be wrong, but it seems they need to pass RNGs since they don't make a distinction between parameters and states? https://flax.readthedocs.io/en/latest/flax.linen.html?highlight=make_rng#flax.linen.Module.apply. In our case, layers should store their own RNG in their state

**Comment by avik-pal** [comment]
[fba8ca2](https://github.com/avik-pal/ExplicitFluxLayers.jl/pull/7/commits/fba8ca24743ec7f3c9e793923a473ff73babf000) changes the default parameterization to NamedTuples from ComponentArrays. ComponentArrays still work but generating a ComponentArray with 90M parameters takes too long (>30 min). For problems requiring an AbstractArray (mostly SciML) the parameterization would not be that large.

File: src/core.jl (hunk 7 above)
[the remark this answers is not in the document]
Comment (ToucheSir) [reply src/core.jl:30]: AFAICT they do, but the internals are quite complex so it's hard to figure out where everything is threaded through. I suppose you could include it as an input if need be, but that seems awkward if `x` is not varargs.

File: src/layers/dropout.jl [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+    @assert 0 ≤ p ≤ 1
+    return Dropout(p, initial_seed, dims)
+end
+
+Base.@pure function (d::Dropout{T})(x::AbstractArray{T}, ps, st::NamedTuple) where {T}
+    !istraining(st) || return (x, st)
+    y = dropout(MersenneTwister(st.seed), x, d.p; dims=d.dims)[1]
+    @set! st.seed = st.seed + 1
+    return y, st
+end
```
[the remark this answers is not in the document]
Comment (avik-pal) [reply src/layers/dropout.jl]: I tried this locally. How do we "run" the underlying RNG?
```julia
julia> zz, st_ = d(x, ps, st);

julia> zz2, st_ = d(x, ps, st_);

julia> zz == zz2
true
```

[the remark this answers is not in the document]
Comment (ToucheSir) [reply src/layers/dropout.jl]: It should update, but I made a typo in my suggestion. Setting `st.rng = rng` should do the trick :)

File: src/layers/basic.jl (hunk 4 above)
[the remark this answers is not in the document]
Comment (avik-pal) [reply src/layers/basic.jl]: The last commit makes things better but still not type inferable
```julia
d = EFL.Dense(2, 2; bias=false)
psd, std = EFL.setup(Random.default_rng(), d) .|> gpu
x = randn(Float32, 2, 1) |> gpu
@code_adjoint d(x, psd, std)
# Zygote.Adjoint(1: (%5, %6 :: Zygote.Context, %1, %2, %3, %4)
#   %7 = Zygote._pullback(%6, Zygote.literal_getproperty, %3, Val{:weight}())
#   %8 = Base.getindex(%7, 1)
#   %9 = Base.getindex(%7, 2)
#   %10 = Zygote._pullback(%6, ExplicitFluxLayers.fast_matmul, %8, %2)
#   %11 = Base.getindex(%10, 1)
#   %12 = Base.getindex(%10, 2)
#   %13 = Zygote._pullback(%6, Core.tuple, %11, %4)
#   %14 = Base.getindex(%13, 1)
#   %15 = Base.getindex(%13, 2)
#   return %14, 1: (%1)
#   %2 = (@15)(%1)
#   %3 = Zygote.gradindex(%2, 2)
#   %4 = Zygote.gradindex(%2, 3)
#   %5 = (@12)(%3)
#   %6 = Zygote.gradindex(%5, 2)
#   %7 = Zygote.gradindex(%5, 3)
#   %8 = (@9)(%6)
#   %9 = Zygote.gradindex(%8, 2)
#   %10 = Zygote.tuple(nothing, %7, %9, %4)
#   return %10)

y, back = pullback(d, x, psd, std)
@code_warntype back.back(y)
# MethodInstance for (::typeof(∂(λ)))(::Tuple{CuArray{Float32, 2, CUDA.Mem.DeviceBuffer}, NamedTuple{(), Tuple{}}})
#     from (j::Zygote.Pullback{T})(Δ) where T in Zygote at compiler/interface2.jl:43
#   Static Parameters
#     T = Tuple{ExplicitFluxLayers.Dense{false, typeof(identity), typeof(Flux.glorot_uniform), typeof(Flux.zeros32)}, CuArray{Float32, 2, CUDA.Mem.DeviceBuffer}, NamedTuple{(:weight,), Tuple{CuArray{Float32, 2, CUDA.Mem.DeviceBuffer}}}, NamedTuple{(), Tuple{}}}
#   Arguments
#     #self#::typeof(∂(λ))
#     Δ::Tuple{CuArray{Float32, 2, CUDA.Mem.DeviceBuffer}, NamedTuple{(), Tuple{}}}
#   Body::Tuple{Nothing, CuArray{Float32, 2, CUDA.Mem.DeviceBuffer}, Union{Nothing, NamedTuple{(:weight,), Tuple{CuArray{Float32, 2, CUDA.Mem.DeviceBuffer}}}}, NamedTuple{(), Tuple{}}}
#   1 ─       $(Expr(:meta, :inline))
#   │   %2  = Base.getfield(#self#, :t)::Tuple{typeof(∂(fast_matmul)), Zygote.var"#1630#back#159"{typeof(identity)}, Zygote.var"#1761#back#222"{Zygote.var"#back#221"{:weight, Zygote.Context, NamedTuple{(:weight,), Tuple{CuArray{Float32, 2, CUDA.Mem.DeviceBuffer}}}, CuArray{Float32, 2, CUDA.Mem.DeviceBuffer}}}}
#   │   %3  = Base.getindex(%2, 3)::Zygote.var"#1761#back#222"{Zygote.var"#back#221"{:weight, Zygote.Context, NamedTuple{(:weight,), Tuple{CuArray{Float32, 2, CUDA.Mem.DeviceBuffer}}}, CuArray{Float32, 2, CUDA.Mem.DeviceBuffer}}}
#   │   %4  = Base.getindex(%2, 2)::Core.Const(Zygote.var"#1630#back#159"{typeof(identity)}(identity))
#   │   %5  = Base.getindex(%2, 1)::typeof(∂(fast_matmul))
#   │   %6  = (%4)(Δ)::Tuple{Nothing, CuArray{Float32, 2, CUDA.Mem.DeviceBuffer}, NamedTuple{(), Tuple{}}}
#   │   %7  = Zygote.gradindex(%6, 2)::CuArray{Float32, 2, CUDA.Mem.DeviceBuffer}
#   │   %8  = Zygote.gradindex(%6, 3)::Core.Const(NamedTuple())
#   │   %9  = (%5)(%7)::Tuple{Nothing, CuArray{Float32, 2, CUDA.Mem.DeviceBuffer}, CuArray{Float32, 2, CUDA.Mem.DeviceBuffer}}
#   │   %10 = Zygote.gradindex(%9, 2)::CuArray{Float32, 2, CUDA.Mem.DeviceBuffer}
#   │   %11 = Zygote.gradindex(%9, 3)::CuArray{Float32, 2, CUDA.Mem.DeviceBuffer}
#   │   %12 = (%3)(%10)::Union{Nothing, Tuple{Nothing, NamedTuple{(:weight,), Tuple{CuArray{Float32, 2, CUDA.Mem.DeviceBuffer}}}, Nothing}}
#   │   %13 = Zygote.gradindex(%12, 2)::Union{Nothing, NamedTuple{(:weight,), Tuple{CuArray{Float32, 2, CUDA.Mem.DeviceBuffer}}}}
#   │   %14 = Zygote.tuple(nothing, %11, %13, %8)::Tuple{Nothing, CuArray{Float32, 2, CUDA.Mem.DeviceBuffer}, Union{Nothing, NamedTuple{(:weight,), Tuple{CuArray{Float32, 2, CUDA.Mem.DeviceBuffer}}}}, NamedTuple{(), Tuple{}}}
#   └──       return %14
```

Though the adjoint code seems more reasonable than what Flux generates

File: src/autodiff.jl [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+# Layers are stateless so we can simply return that
+Base.zero(l::AbstractExplicitLayer) = l
+
+ChainRulesCore.rrule(::typeof(istraining)) = true, _ -> (NoTangent(),)
+function ChainRulesCore.rrule(::typeof(istraining), st::NamedTuple)
+    return (st.training == :auto ? true : st.training), _ -> (NoTangent(), NoTangent())
+end
```
[the remark this answers is not in the document]
Comment (avik-pal) [reply src/autodiff.jl]: I have changed this to ensure that `training` is only Boolean (at least for now).

File: src/layers/dropout.jl [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+    return VariationalHiddenDropout(p, initial_seed, dims)
+end
+
+Base.@pure function (d::VariationalHiddenDropout{T})(
+    x::AbstractArray{T}, ps, st::NamedTuple
+) where {T}
+    !istraining(st) || return (x, st)
+    if st.update_mask
+        y, mask = dropout(MersenneTwister(st.seed), x, d.p; dims=d.dims)
+        return y, (mask=mask, seed=st.seed + 1, update_mask=false, training=st.training)
```
[the remark this answers is not in the document]
Comment (avik-pal) [reply src/layers/dropout.jl]: Yes, we don't know the input size at model creation time

**Comment by avik-pal** [comment]
We are now quite close to merging. A few things I am working on rn:
- [ ] Remove the Flux dependency
- [ ] Remove the Yota.jl adjoints
- [ ] Add an example imagenet training script like flax/pytorch

**Comment by ToucheSir** [comment]
If you see anything missing from NNlib that should be in NNlib (e.g. functional versions of norm layers), don't hesitate to drop a PR and ping me.
